### DISCUSSION TITLE
Version without (c)chardet?

### DISCUSSION THREAD

**Discussion by EdLeafe** [aio-libs/aiohttp#5888] [General]
I really like using aiohttp for the app I'm developing for work, but the legal team has raised a red flag regarding the use of `chardet` (or `cchardet`), as it is LGPL-licensed. I've looked through the source, and see that its only usage is in the `get_encoding()` method of `client_reqrep.py`. I tried removing that `else:` block, and everything runs fine, as we are not processing unknown blocks of text. I saw that you made it a required dependency [back in 2015](https://github.com/aio-libs/aiohttp/issues/318), but the license for it makes the entire aiohttp project not FOSS-compatible.

**Comment by Dreamsorcerer** [comment]
Maybe it's a good idea to make it work without it, but I'm not sure what the issue is with a LGPL library or how that makes it "not FOSS-compatible" (it's literally a FOSS license).

The only reason I can think of that you would even need to think about doing something extra to comply with the LPGL is if you release a Python app as a complete binary blob including the library, in which case you would need to have a method to allow a user to replace the library in the binary (but you could also just tweak your build process to exclude the library, either by patching aiohttp or by including a dummy version of the library). In the typical case for Python software, where the library is just depended on and pip installed, you would not need to do anything.

**Reply by EdLeafe** [reply]
Sorry, I meant to write "not FOSS-compatible because we need to release a binary blob". Well, actually, it's the binary blob that is the problem. We have such a substitution method, but it's a bit of a hack, and it would be simpler to just remove chardet, and the legal team here at BIG COMPANY is still uneasy about it.

**Reply by Dreamsorcerer** [reply]
I'm not sure what other maintainer's thoughts on this are, but my feeling is that the build process can be used to exclude the library. But, maybe we could introduce a try/except thing to not require importing it.

As an immediate solution though, I think the simplest thing to do is to have a custom version of chardet that gets included into the binary instead, which could literally just be an empty `__init__.py` file. Or, to be a little more thorough:

```
def detect(*args, **kwargs):
    raise NotImplementedError("Real chardet library not installed")
```

**Reply by Ousret** [reply]
Hi there,

Recently, `requests` maintainers did help toward this. Charset-Normalizer is a viable alternative (MIT licensed). Could provide a smooth transition.
Happy to help.

Regards,

**Reply by EdLeafe** [reply]
That looks promising - thanks, @Ousret !

**Comment by Ousret** [comment]
Hi,

In case someone stumbles here, I have proposed a PR for that matter.
https://github.com/aio-libs/aiohttp/pull/5930

**Reply by asvetlov** [reply]
Done by aiohttp 3.8.0

**Comment by EdLeafe** [comment]
Great! Thanks to all for a great product!

**Reply by asvetlov** [reply]
Thank you!
